### ISSUE TITLE
Image+ outputs full server path for [[+url]]

### ISSUE THREAD

**Issue by hexcollective** [Jako/ImagePlus#41]
Hi,
Image+ is outputting [[+url]] value as full server path, i.e. src="/var/www/vhosts/abc.co.uk/123.abc.co.uk/assets/uploads/images/roadmap.jpeg"
The intended path would be: /assets/uploads/images/roadmap.jpeg
Hence the image shows as broken.

The Media source for the Image+ TV is > basePath + baseUrl: assets/uploads/ 

Any ideas?

**Comment by Jako** [comment]
`basePath` should contain the whole path to `assets/uploads/`. Maybe thats the problem.

**Comment by Jako** [comment]
But it looks more like a bug.

**Comment by Parthian** [comment]
I've just been investigating this same problem. Didn't notice this. 
I'm sure this is a bug as you say and one that has security implications. This may help. I've redacted it.

Revo 2.5.0.  ImagePlus 2.4.5
3 ImagePlus TVs with Output as ImagePlus and the default field set to paths like this.
assets/img/superwide/grinder-closeup.jpg
Also a Media Source is set with the following 
 
basePath   assets/img/
basePathRelative    Yes
baseUrl assets/img/
baseUrlRelative    Yes

Chunk has this.
                 <img class="col-sm-4 img-responsive" src="[[*tvImageTriple1]]" alt="Overview 1" />
                 <img class="col-sm-4 img-responsive" src="[[*tvImageTriple2]]" alt="Overview 2" />
                 <img class="col-sm-4 img-responsive" src="[[*tvImageTriple3]]" alt="Overview 3" />

I edit the resource with the TVs such that only the first TV has content. Expected the others to use the default. Instead I get this HTML

<img class="col-sm-4 img-responsive" src="/assets/components/phpthumbof/cache/pressure-vessel-manufacture.c2000723fb96f9507f9e952c5e9a7dff.jpg" alt="Overview 1">
<img class="col-sm-4 img-responsive" src="/home/MYUSERNAME/SERVERPATH/PROJECTFOLDER/assets/img/assets/img/superwide/grinder-closeup.jpg" alt="Overview 2">
<img class="col-sm-4 img-responsive" src="/home/MYUSERNAME/SERVERPATH/PROJECTFOLDER/assets/img/assets/img/superwide/mill-electrical-controls.jpg" alt="Overview 3">

Information in bold would be really useful to a hacker. And also note that the /assets/img/assets/img may simply be an error on my part due to the Media Source already pointing to that image folder (update to the docs to clarfy what a default should be???). 

Also, it gets odder in the Error Log. (just one of the two missing images shown)
/home/MYUSERNAME/SERVERPATH/PROJECTFOLDER/core/components/phpthumbof/model/phpthumbof.class.php : 156) [pThumb] Resource: 60 || Image: (none)
File not found: /home/MYUSERNAME/SERVERPATH/PROJECTFOLDER/home/MYUSERNAME/SERVERPATH/PROJECTFOLDER/assets/img/assets/img/superwide/grinder-closeup.jpg  *** Skipping ***

Key point being the repitition in the file not found line.

Tried debugging it but no progress.
I've wondered about the imageplus snippet "check" option with this bug. Perhaps the corepath??? error will incorrectly be interpreted as "image" as opposed to "no image". 

Hope that helps and that the code formatting works. 
Stuart

**Comment by hexcollective** [comment]
My media source set up is also Relative for : basePath + baseUrl

**Comment by Jako** [comment]
I can't reproduce the issues here.

@hexcollective Is the Media Source assigned in the TV? Have you reassigned the image after assigning the Media Source?

@Parthian The default is set wrong. The basepath of the Media Source has to be removed from the  default. It has to be `superwide/grinder-closeup.jpg`. BTW: You could format your answer with Markdown and surround the code parts with three backticks. There is also a toolbar above the edit window.

**Comment by Jako** [comment]
Found an issue: If the Image+ TV is reset to no value, the core path is shown.

**Comment by hexcollective** [comment]
Hi @Jako,

> Is the Media Source assigned in the TV?

 Yes, the media source is set as default and is assigned to the Image+ TV.

> Have you reassigned the image after assigning the Media Source? 

Not 100% sure what you mean here though I just deleted the original TV and created a new one, assigned Image+ to input + output, media source (assets/uploads/) is applied. The url still displays as full server path (src="/var/www/vhosts/abc.co.uk/123.abc.co.uk/assets/uploads/images/roadmap.jpeg")

**Comment by Jako** [comment]
The MODX error log should show something like this: `The template variabe value does not contain an existing image`

**Comment by hexcollective** [comment]
Hi @Jako 
I didn't see that error though I did note a resizer issue on pthumb (class not loaded) so I have disabled this and the image is now rendering!

**Comment by Jako** [comment]
It was just a INFO debug message, so it maybe did not occur in your log.

**Comment by Parthian** [comment]
Hi Jako,
I've used your new code. Presumably the json='' clears the corepath if a TV is empty or the default path is wrong.

Works well - I can't see the innards of my server when images are missing. 

But oddly I've kept the Input default at assets/img/superwide/grinder-closeup.jpg  and cleared the 3rd TV (to simulate an editing error or just a desire not to have a 3rd image) and the default appeared perfectly. Which is contrary to what your commented suggested. 

> The default is set wrong. The basepath of the Media Source has to be removed from the default. 

Changing the default to remove the Media Source part assets/img/  results in no image, just src="superwide/mill-electrical-controls.jpg" and my Alt text is displayed (pre fix this would have been the corepath stuff).

It looks like the Media Source is ignored in building the path to the default image. A  new bug? Or perhaps hidden by the corepath issue?

Thanks
Stuart

**Comment by Jako** [comment]
The internal value of an Image+ TV is not just a path to the image. It contains also the used Media Source, the cropping etc. Please try out what default value of the TV leads to a suitable result.

**Comment by Parthian** [comment]
Possible language barrier. I'm not sure if you are suggesting - "just use what works". Or if you are unsure of the problem. Assuming the latter. 

Media Source = assets/img
ImagePlus Default Example = afilename.jpg 

I would assume that internally, ImagePlus would append the default to the media source. This is what happens when the Resource TV is updated - it doesn't have the full assets/img/ path. Which is fine.

Problem. If I leave the imageplus TV empty then I want the default to jump in. But the image does not appear because the default is "afilename.jpg" when the HTML img src needs "assets/img/afilename.jpg"

So the expected Append isn't happening. This isn't the end of the world, I just need to add assets/img/ to the filename of the image I want as a default but it is inconsistent and possibly a bug as it contradicts your earlier statement  

> The default is set wrong. The basepath of the Media Source has to be removed from the default. It has to be superwide/grinder-closeup.jpg
I need to set the default to assets/img/superwide/grinder-closeup,jpg  for it to appear in the HTML.

Apologies for labouring the point.

**Comment by Jako** [comment]
If the (default) value of a Image+ TV is only a path, then Image+ tries to transform this path into an real Image+ TV value. Maybe the src path of that Image+ TV value is not valid because of the Media Source value. But if it is found, that is not a real issue.

**Comment by Parthian** [comment]
Okay, it sounds like a 'feature'. Easily worked around. And should be fairly obvious what's gone missing for a typical developer. Perhaps consider updating the docs. 

Enjoying setting the ratio for images. Really helps. Cheers.

**Comment by Jako** [comment]
@Parthian I have patched the wrong file before, now a not found file leads to show the invalid tv value instead of feeding the cropping engine.

**Comment by Parthian** [comment]
I've updated the 3 files. Don't see any change to operation. What should I see?
Thinking.
1) Default, if used, should be a PHPThumb Cached version? This isn't happening.
OR
2) Your fix was for the Media Source to be taken into account. This isn't happening. Same as before, the default in the Input part of the TV has to be assets/img/blah.jpg
OR
Something else. I have noticed that if I try to Save a Resource when the TV is set to a non existing file the Save is not permitted. Which is very clever, but may have been like that before.

Summary
Not sure what to test for. Do I need to clear the core/cache folder when making code updates (used the Manager Clear Cache)?

**Comment by Jako** [comment]
The Image+ TV has used the Same media Source to locate the image in all three cases. There should be some small difference comparing the latest patches when the image path was not valid.

Location | Before | Patch 1 | Patch 2 
---------|--------|--------|---------
Frontend output | `[[+url]]` contained the full server path | `[[+url]]` contains no value | `[[+url]]` contains the tv src value
Backend value | Image input of the TV contains the invalid path | Image input of the TV contains no value | Image input of the TV contains the invalid path

**Comment by Jako** [comment]
In your case `superwide/grinder-closeup.jpg` should be valid if the Media Source contains the following values:

Setting | Value
--------|------
basePath | `assets/img/`
basePathRelative | Yes
baseUrl | `assets/img/`
baseUrlRelative | Yes

**Comment by Parthian** [comment]
Apologies for the delay. Missed the email.
I confirm my media source settings are as you state.
But I need assets/img/superwide/grinder-closeup.jpg  in the TV Input Default field for the image to appear. 

Therefore on that basis there is a bug. Perhaps I have some other issue? Maybe @hexcollective can try and replicate it.

**Comment by Jako** [comment]
@Parthian I have checked that multiple on my dev installation. An image is found and rendered in the front- and backend with a Media Source containing the paths above and having only a filename without a path in the file input or in the default value.

Maybe it is a slash issue in your installation. The paths have to use a trailing `/` and the image should not start with one.

**Comment by Parthian** [comment]
Morning Jako,
I always have trouble using MODx with Friendly URLs and usually have trouble with image paths. Often I develop on a subdomain and then move to the 'proper' domain and images break. 

So I looked at my FURLs settings. I have "Use Friendly Alias Paths - use_alias_path"  set to Yes. Which I suspect is common e.g. domain.com/parent/child.html  as opposed to domain.com/child.html
This has a warning:

> NOTE: When setting this option to 'Yes' (turning on alias paths), reference items (such as images, CSS, JavaScripts, etc.) use the absolute path, e.g., '/assets/images' as opposed to 'assets/images'. By doing so you will prevent the browser (or web server) from appending the relative path to the alias path.  

Oddly, throughout my header templates etc. I have things like   assets/css/somestyle.css   - no prepended /. Which is technically incorrect given the warning. So I tried changing my media source paths by prepending the slash   from assets/  to  /assets/.. 

Setting 	Value
basePath 	/assets/img/
baseUrl 	/assets/img/

And the "path/image.jpg" default approach now works as expected. Inspection of the page shows this - src="/assets/components/phpthumbof/cache/grinder-closeup.ffb...etc.jpg". Whereas when I forced the default to "work" by using assets/img/superwide/grinder-closeup.jpg that is exactly what I got in the src i.e. the phpthumbof cache wasn't used. So all now works. 

And amazingly I haven't broken the rest of the sites images... so it looks like I should change the css/javascript links in templates etc. by using /assets... for robustness. Hopefully it will all work when moving to the production setup.

Thanks for your patience. But between Hex and myself I think we've found a bug or two. 

Stuart

**Comment by mdillgen** [comment]
Hey,

This is closed but Is there any solution yet. I still got the same problem of image+ displaying full server path in the frontend! I really can't resolve it.

Marcus

**Comment by hexcollective** [comment]
Hello and thanks for your email. I'm out of the office today and will respond to your e-mail as soon as possible.

If you require hosting or website retainer support, please send an email with the details to: support@hexcollective.co.uk and/or text: 07861 297 630.

Kind regards,

Alistair

Hex Collective Ltd

**Comment by lawebfabric** [comment]
Hello,

I got the same too, all test give the same the full server path.
Did someone found a solution ?

**Comment by hexcollective** [comment]
Hello and thanks for your email. I'm away on holiday and will respond to your e-mail as soon as possible.

If you require hosting support, please visit this link and send a description of the issue: https://www.hexcollective.co.uk/support/

Kind regards,

Alistair

Hex Collective Ltd

**Comment by Grapestain** [comment]
Hi!

I know I'm late to the party, but I'm facing the same issue! Or maybe the same, not sure.

I'm trying to - in some cases - get around the whole resizing and get the URL of the original source image. I make this selectable using another TV, so I cannot just use a regular image TV.

My attempt to do so was to use this:
```[[ImagePlus? &value=`[[+icon]]` &type=`tpl` &tpl=`ImagePlus.source.src`]]```

where `[[+icon]]` is my ImagePlus TV, and `ImagePlus.source.src` chunk contains this:
`[[+source.src]]`

Based on the [documentation](https://jako.github.io/ImagePlus/usage/#output-chunk), this should result in the "URL of the source image" being returned. However what I got is the server file path of the original image - which indeed I cannot use in the `src` attribute of my `img` tag - which is my problem.

I've looked into it a bit, and I've noticed that in line 160 of`core/components/imageplus/model/cropengines/PhpThumbOf.php` really the `$imgPath` variable is being assigned to the source.src placeholder, which is in fact a server path, and not an URL:
```
            $chunkParams = array_merge($opts, array(
                'url' => $url,
                'alt' => $data->altTag,
                'width' => $data->targetWidth,
                'height' => $data->targetHeight,
                'source.src' => $imgPath,
                'source.width' => $data->sourceImg->width,
                'source.height' => $data->sourceImg->height,
                'crop.width' => $data->crop->width,
                'crop.height' => $data->crop->height,
                'crop.x' => $data->crop->x,
                'crop.y' => $data->crop->y,
                'options' => $options,
                'crop.options' => $cropOptions,
                'caption' => isset($data->caption) ? $data->caption : '',
                'credits' => isset($data->credits) ? $data->credits : ''
            ));
            return $this->modx->getChunk($outputChunk, $chunkParams);
```

Long story short: do you think it is the way it should be and I'm misinterpreting the documentation, or is it really a bug and an URL should be assigned instead of a path?

**Comment by Jako** [comment]
pThumb works with a full server path or with a relative base url based path, if I remember right.

**Comment by Grapestain** [comment]
Most probably that's true. pThumb works fine indeed.

What I find confusing is that the documentation suggest one can use `[[+source.src]]` in a template to access the URL of the original image, but instead it contains the path of it. Seems to be contradictory to me... ?

**Comment by Jako** [comment]
Ok, the docs should be changed. Could you create a PR with your suggestions?
